api/src/pcapi/core/token/__init__.py

```diff
@@ -0,0 +1,87 @@
+import jwt
+
+from pcapi.core.users import utils
+from pcapi.core.users.exceptions import InvalidToken
+
+
+class TokenType(enum.Enum):
+    EMAIL_CHANGE_CONFIRMATION = "update_email_confirmation"
+    EMAIL_CHANGE_VALIDATION = "update_email_validation"
+
+
+@dataclasses.dataclass(frozen=True)
+class Token:
+    type_: TokenType
+    user_id: int
+    encoded_token: str
+    data: dict
+
+    @classmethod
+    def load(cls: "Token", encoded_token: str) -> "Token":
+        try:
+            payload = utils.decode_jwt_token(encoded_token)
+        except jwt.PyJWTError:
+            raise InvalidToken()
+        try:
+            data = payload["data"] if payload["data"] is not None else {}
+            type_ = TokenType(payload["token_type"])
+            user_id = payload["user_id"]
+            return cls(type_, user_id, encoded_token, data)
+        except (KeyError, ValueError):
+            raise InvalidToken()
+
+    @classmethod
+    def _get_redis_key(cls, type_: TokenType, user_id: int) -> str:
+        return f"pcapi:token:{type_.value}_{user_id}"
+
+    def get_expiration_date(self) -> datetime | None:
+        ttl = app.redis_client.ttl(Token._get_redis_key(self.type_, self.user_id))  # type: ignore [attr-defined]
+        if ttl < 0:
+            # -2 if doesn't exist, -1 if no expiration
+            return None
+        return datetime.utcnow() + timedelta(seconds=ttl)
+
+    def check(self, type_: TokenType, user_id: int | None = None) -> None:
+        if (
+            not app.redis_client.exists(Token._get_redis_key(self.type_, self.user_id))  # type: ignore [attr-defined]
+            or app.redis_client.get(Token._get_redis_key(self.type_, self.user_id)) != self.encoded_token  # type: ignore [attr-defined]
+            or self.type_ != type_
+            or (user_id is not None and self.user_id != user_id)
+        ):
+            raise InvalidToken()
+
+    def expire(self) -> None:
+        app.redis_client.delete(Token._get_redis_key(self.type_, self.user_id))  # type: ignore [attr-defined]
+
+    @classmethod
+    def load_and_check(cls: "Token", encoded_token: str, type_: TokenType, user_id: int) -> "Token":
+        token = Token.load(encoded_token)
+        token.check(type_, user_id)
+        return token
+
+    @classmethod
+    def create(
+        cls: "Token", type_: TokenType, ttl: timedelta | None, user_id: int, data: dict | None = None
+    ) -> "Token":
+        encoded_token = utils.encode_jwt_payload({"token_type": type_.value, "user_id": user_id, "data": data})
+        app.redis_client.set(cls._get_redis_key(type_, user_id), encoded_token, ex=ttl)  # type: ignore [attr-defined]
+        return Token.load(encoded_token)
+
+    @classmethod
+    def token_exists(cls, type_: TokenType, user_id: int) -> bool:
+        return app.redis_client.exists(cls._get_redis_key(type_, user_id))  # type: ignore [attr-defined]
+
+    @classmethod
+    def get_token(cls: Type[T], type_: TokenType, user_id: int) -> T:
```

Comment:
<!-- Issues represent user-facing problems. If possible, it's great to follow this kind of comment with a suggestion.  -->
**🔨 issue:** ‏Oubli d'un `T` ici